firmware/thruna/FSM/Src/Io/Io_AcceleratorPedals.c

```diff
@@ -1,76 +1,125 @@
+
+#define SAPPS_COS_LAW_DENOMINATOR (2.0f * SAPPS_LEN_A * LEN_B)
+#define SAPPS_COS_LAW_COEFFICIENT ((SAPPS_LEN_A * SAPPS_LEN_A + LEN_B * LEN_B) / SAPPS_COS_LAW_DENOMINATOR)
+
+// Macro to calculate the linear change in PAPPS length per change in voltage (we only use max because min = 0)
+#define APPS_VOLTAGE_PER_MM ((MAX_POT_LENGTH - MIN_POT_LENGTH) / MAX_POT_V)
+
+// linear equation to calculate current length using a rate of change from potientiometer
+// here max length here is the Yint and apps voltager per mm is the slope
+#define RAW_VOLTAGE_TO_LEN_MM(voltage) (MAX_POT_LENGTH - voltage * APPS_VOLTAGE_PER_MM)
 
-#define MIN_PEDAL_VOLTAGE (0.5f)
-#define MAX_PEDAL_VOLTAGE (3.1f)
+// calculating the unpressed and pressed PAPPS/SAPPS length based on the measured pot readings
+#define PAPPS_LENGTH_UNPRESSED_MM (RAW_VOLTAGE_TO_LEN_MM(PAPPS_UNPRESSED_POT_V))
+#define PAPPS_LENGTH_FULLY_PRESSED_MM (RAW_VOLTAGE_TO_LEN_MM(PAPPS_FULL_PRESSED_POT_V))
 
-// NONLINEAR APPROXIMATION
-#define FULLY_UNPRESSED_POT_LENGTH_MM (212.2f)
-#define FULLY_PRESSED_POT_LENGTH_MM (165.6f)
+#define SAPPS_LENGTH_UNPRESSED_MM (RAW_VOLTAGE_TO_LEN_MM(SAPPS_UNPRESSED_POT_V))
+#define SAPPS_LENGTH_FULLY_PRESSED_MM (RAW_VOLTAGE_TO_LEN_MM(SAPPS_FULL_PRESSED_POT_V))
 
-#define PAPPS_DEADZONE_PERCENT (10.0f)
-#define SAPPS_DEADZONE_PERCENT (8.0f)
+// OC & SC bounds
+#define PAPPS_MIN_V (PAPPS_UNPRESSED_POT_V - 0.5f)
+#define PAPPS_MAX_V (PAPPS_FULL_PRESSED_POT_V + 0.5f)
+
+#define SAPPS_MIN_V (SAPPS_UNPRESSED_POT_V - 0.5f)
+#define SAPPS_MAX_V (SAPPS_FULL_PRESSED_POT_V + 0.5f)
+
+#define DEAD_ZONE_PERCENT (10.0f)
+
+static float calcAppsAngle(float cos_law_coefficent, float cos_law_denominator, float pot_len)
+{
+    float angle = (acosf(cos_law_coefficent - (powf(pot_len, 2) / cos_law_denominator)));
+
+    return angle;
+}
 
 float Io_AcceleratorPedals_GetPapps(void)
 {
-    // Length calculated via voltage
-    float       pedal_voltage = Io_Adc_GetChannelVoltage(ADC1_CHANNEL_2);
-    const float pot_len_mm    = PAPPS_RAW_VOLTAGE_TO_LEN_MM(pedal_voltage);
+    float pedal_voltage = Io_Adc_GetChannelVoltage(ADC1_CHANNEL_2);
+
+    // Length calculated via voltage reading from PAPPS
+    const float pot_len_mm = RAW_VOLTAGE_TO_LEN_MM(pedal_voltage);
+
+    const float rest_angle =
+        calcAppsAngle(PAPPS_COS_LAW_COEFFICIENT, PAPPS_COS_LAW_DENOMINATOR, PAPPS_LENGTH_UNPRESSED_MM);
 
-    // Calculate pedal percentage
-    float pedal_percentage = (pot_len_mm - FULLY_UNPRESSED_POT_LENGTH_MM) /
-                                 (FULLY_PRESSED_POT_LENGTH_MM - FULLY_UNPRESSED_POT_LENGTH_MM) * 100.0f -
-                             PAPPS_DEADZONE_PERCENT;
+    float pedal_angle = rest_angle - calcAppsAngle(PAPPS_COS_LAW_COEFFICIENT, PAPPS_COS_LAW_DENOMINATOR, pot_len_mm);
+
+    const float max_angle =
+        rest_angle - calcAppsAngle(PAPPS_COS_LAW_COEFFICIENT, PAPPS_COS_LAW_DENOMINATOR, PAPPS_LENGTH_FULLY_PRESSED_MM);
```

Comment:
`rest_angle` and `max_angle` are constants since they're just a pure function of `#define`s, so we theoretically only need to calculate them once, instead of every time this function is called. If you define them as static variables in the global scope, you can make an `Io_AcceleratorPedals_Init` function that calculates their values. Then just remember to call this init function from `main.c`. This is probably worth it since we know this is an expensive operation (was causing watchdog timeouts before we increased the clock speed)